libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/targets/android/Android25ProjectHandler.kt

```diff
@@ -29,99 +26,61 @@ class Android25ProjectHandler(
     kotlinConfigurationTools: KotlinConfigurationTools
 ) : AbstractAndroidProjectHandler(kotlinConfigurationTools) {
 
-    override fun forEachVariant(project: Project, action: (BaseVariant) -> Unit) {
-        val androidExtension = project.extensions.getByName("android")
-        when (androidExtension) {
-            is AppExtension -> androidExtension.applicationVariants.all(action)
-            is LibraryExtension -> {
-                androidExtension.libraryVariants.all(action)
-                if (androidExtension is FeatureExtension) {
-                    androidExtension.featureVariants.all(action)
-                }
-            }
-            is TestExtension -> androidExtension.applicationVariants.all(action)
-        }
-        if (androidExtension is TestedExtension) {
-            androidExtension.testVariants.all(action)
-            androidExtension.unitTestVariants.all(action)
-        }
-    }
-
     override fun wireKotlinTasks(
         project: Project,
         compilation: KotlinJvmAndroidCompilation,
         androidPlugin: BasePlugin,
         androidExt: BaseExtension,
         variantData: BaseVariant,
-        javaTask: AbstractCompile,
-        kotlinTask: KotlinCompile
+        javaTask: TaskProvider<out AbstractCompile>,
+        kotlinTask: TaskProvider<out KotlinCompile>
     ) {
-        val preJavaKotlinOutputFiles = mutableListOf<File>().apply {
-            add(kotlinTask.destinationDir)
-            if (Kapt3GradleSubplugin.isEnabled(project)) {
-                // Add Kapt3 output as well, since there's no SyncOutputTask with the new API
-                val kaptClasssesDir = Kapt3GradleSubplugin.getKaptGeneratedClassesDir(project, getVariantName(variantData))
-                add(kaptClasssesDir)
+        val preJavaKotlinOutput = project.files(project.provider {
+            mutableListOf<File>().apply {
+                add(kotlinTask.get().destinationDir)
```

Comment:
It's hard to see if this is called at the right time, aren't we configuring variants here ? It would be too early, `.get` should only be called during execution time, or when we are sure that the task is created/gonna be created (when you depend on it).